docs/src/tutorial.rst

```diff
@@ -0,0 +1,487 @@
+
+.. _adding_functions:
+
+Step 6: Add the Individual Widget Functions
+===========================================
+
+The next step is to create |LAM| functions to create, show, and delete
+individual widgets.
+Replace the existing **exports.main** function in *widgets.js* with the following code.
+
+.. code-block:: js
+
+    exports.main = async function(event, context) {
+      try {
+        var method = event.httpMethod;
+
+        if (method == "GET") {
+          if (event.path.length == 1) {
+            const data = await S3.listObjectsV2({ Bucket: bucketName }).promise();
+            var body = {
+              widgets: data.Contents.map(function(e) { return e.Key })
+            };
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }
+          else {
+            // GET /name
+            var name = event.path.substring(1, event.path.length);
+  
+            const data = await S3.getObject({ Bucket: bucketName, Key: name}).promise();
+            var body = data.Body.toString('utf-8');
+
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }
+        }
+        else if (method == "POST") {
+          // POST /name
+          // Make sure we have a name
+          if (event.path.length == 1) {
+            // Return error
+            var body = "NAME arg missing (/ supplied, not /name)";
+
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }  
+          else {
+            var name = event.path.substring(1, event.path.length);
+
+            // Create some dummy data to populate object
+            const now = new Date();
+            var data = name + " created: " + now;
+
+            var base64data = new Buffer(data, 'binary');
+
+            await S3.putObject({
+              Bucket: bucketName,
+              Key: name,
+              Body: base64data,
+              ContentType: 'application/json'
+            }).promise();
+
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(event.widgets)
+            };
+          }
+        }
+        else if (method == "DELETE") {
+          // DELETE /name
+          // Make sure we have a name
+          if (event.path.length == 1) {
+            // Return error
+            var body = "NAME arg missing (/ supplied, not /name)";
+        
+            return {
+              statusCode: 200,
+              headers: {},
+              body: JSON.stringify(body)
+            };
+          }
+          else {
+            var name = event.path.substring(1, event.path.length);
+
+            await S3.deleteObject({ 
+              Bucket: bucketName, Key: name
+            }).promise();
+
+            var body = { success: true }
```

Comment:
So in case of success, you're returning an object with `{ success: true }` in it, but in case of failure you're returning a string?

Such an irregular API shape is a pain to work with. Better to return the following then in case of error:

```ts
{ 
  success: false, 
  error: 'This is my error message'
}
```

Reply:
Good idea, I'll make the change(s).